swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -47,295 +72,370 @@ public class Adapter {
   /// - Parameters:
   ///   - completionHandler: completion handler.
   public func start(completionHandler: @escaping (AdapterError?) -> Void) throws {
-    workQueue.async {
-      guard case .stopped = self.state else {
+    workQueue.async { [weak self] in
+      guard let self = self else { return }
+
+      self.logger.debug("Adapter.start")
+      guard case .stoppedTunnel = self.state else {
         completionHandler(.invalidState)
         return
       }
 
-      let networkMonitor = NWPathMonitor()
-      networkMonitor.pathUpdateHandler = { [weak self] path in
-        self?.didReceivePathUpdate(path: path)
-      }
-      networkMonitor.start(queue: self.workQueue)
+      self.callbackHandler.delegate = self
 
+      self.logger.debug("Adapter.start: Starting connlib")
       do {
-        try self.setNetworkSettings(self.generateNetworkSettings(ipv4Routes: [], ipv6Routes: []))
-        self.state = .started(
-          try WrappedSession.connect("http://localhost:4568", "test-token", self.callbackHandler)
+        self.state = .startingTunnel(
+          session: try WrappedSession.connect(self.portalURLString, self.token, self.callbackHandler),
+          onStarted: completionHandler
         )
-        self.networkMonitor = networkMonitor
-        completionHandler(nil)
-      } catch let error as AdapterError {
-        networkMonitor.cancel()
-        completionHandler(error)
-      } catch {
-        fatalError()
+      } catch let error {
+        self.logger.error("Adapter.start: Error: \(error, privacy: .public)")
+        self.state = .stoppedTunnel
+        completionHandler(AdapterError.connlibConnectError(error))
       }
+
     }
   }
 
-  public func stop(completionHandler: @escaping (AdapterError?) -> Void) {
-    workQueue.async {
-      switch self.state {
-      case .started(let wrappedSession):
-        wrappedSession.disconnect()
-      case .temporaryShutdown:
-        break
+  /// Stop the tunnel
+  public func stop(completionHandler: @escaping () -> Void) {
+    workQueue.async { [weak self] in
+      guard let self = self else { return }
 
-      case .stopped:
-        completionHandler(.invalidState)
-        return
+      self.logger.debug("Adapter.stop")
+
+      switch self.state {
+        case .stoppedTunnel, .stoppingTunnel:
+          break
+        case .tunnelReady(let session):
+          self.logger.debug("Adapter.stop: Shutting down connlib")
+          self.state = .stoppingTunnel(session: session, onStopped: completionHandler)
+          _ = session.disconnect()
+        case .startingTunnel(let session, let onStarted):
+          self.logger.debug("Adapter.stop: Shutting down connlib before tunnel ready")
+          self.state = .stoppingTunnel(session: session, onStopped: {
+            onStarted?(AdapterError.stoppedByRequestWhileStarting)
+            completionHandler()
+          })
+          // FIXME: Is it ok to call disconnect() while we haven't got onTunnelReady?
```

Comment:
I can't think of any reason we wouldn't want to allow this, so sounds good to me!